src/compiler/checker.ts

```diff
@@ -27599,7 +27599,11 @@ namespace ts {
                 // If no inferences have been made, and none of the type parameters for which we are inferring
                 // specify default types, nothing is gained from instantiating as type parameters would just be
                 // replaced with their constraints similar to the apparent type.
-                if (inferenceContext && contextFlags! & ContextFlags.Signature && some(inferenceContext.inferences, hasInferenceCandidatesOrDefault)) {
+                if (
+                    inferenceContext &&
+                    (contextFlags! & ContextFlags.Signature || maybeTypeOfKind(contextualType, TypeFlags.Conditional) || maybeTypeOfKind(contextualType, TypeFlags.IndexedAccess)) &&

```

Comment:
I've only included conditional types and indexed access types here but perhaps some other types could be added here too. The idea here is that those types are often used to "route" the type/select one of multiple options somehow. And thus those are worth checking for here. 

Could the same be said about any other type? 🤔

Reply:
I've discovered there is already `TypeFlags.Simplifiable` that includes `TypeFlags.IndexedAccess` and `TypeFlags.Conditional`. So I conclude that this is exactly what I want here and that no other types should be included here.